defopt.py

```diff
@@ -104,20 +106,32 @@ def _populate_parser(func, parser, parsers, short):
     doc = _parse_doc(func)
     hints = _get_type_hints(func)
     parser.description = doc.text
+
+    types = {name: _get_type(func, name, doc, hints)
+             for name, param in sig.parameters.items()}
+    positionals = {name for name, param in sig.parameters.items()
+                   if (param.default == param.empty
+                           and not types[name].container
+                           and param.kind != param.KEYWORD_ONLY)}
+    if short is None:
+        count_initials = Counter(name[0] for name in sig.parameters
+                                 if name not in positionals)
+        short = {name.replace('_', '-'): name[0] for name in sig.parameters
+                 if name not in positionals and count_initials[name[0]] == 1}
```

Comment:
I would prefer to use single letter variable names in list and dictionary comprehensions. This code works fine currently, but the difference in behavior between Python 2 and 3 has caused me some grief in the past. (The variables remain in the current scope in 2, but not in 3).